drivers: led_strip: Various enhancements

thedjnK opened the pull request:
Various enhancements to the LED strip driver class:

- Adds a common binding with (mandatory) chain length and colour ordering properties
- Adds a length function which returns the length of the LED strip
- Fixes some formatting issues with the driver
- Check length before updating LED strip
- Make update channels function optional
- Add chain length and colour mapping to LED strips in boards and overlays
- Remove LED length Kconfig from sample as chain length is now mandatory

thedjnK committed:
dts: bindings: led_strip: Add common binding

Adds a common binding with a chain length and colour ordering
property

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
drivers: led_strip: Add length function

Adds a length function which returns the length of the LED strip

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
drivers: led_strip: Update format

Updates the include file to fix some formatting issues

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
drivers: led_strip: Check length before updating LED strip

Checks that the supplied length is valid for the given driver
before passing it to the update function

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
drivers: led_strip: Make update channels function optional

Makes the update channels function optional, this is only
implemented in one driver so can be safely omitted from most
drivers

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
boards: Add chain length and colour mapping to LED strips

Adds chain length and colour mapping properties to LED strip devices
on boards where these parameters are now required

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
doc: migration/release: 3.7: Add LED strip changes

Adds changes about the LED strip driver

Signed-off-by: Jamie McCrae <spam@helper3000.net>

soburi commented on include/zephyr/drivers/led_strip.h:
```diff
@@ -71,6 +72,14 @@ typedef int (*led_api_update_channels)(const struct device *dev,
+ *
+ * @see led_strip_length() for argument descriptions.
+ */
+typedef int (*led_api_length)(const struct device *dev);
```
This value is quantity. So, It should be better unsigned value.

kartben replied:
Reason of it being an int is that it may return a negative errno

thedjnK replied:
As in, always return the length even if the driver init failed and never return an error from this function?

soburi replied:
There are implementation examples for the counter API that simply returns a fixed value without returning an error.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/counter.h#L294-L302

Similar to these, we expect an implementation that simply returns the value of devicetree as is, so I think it would be a good idea to follow this example.

thedjnK replied:
Fixed

thedjnK committed:
samples: drivers: led_strip: Remove length Kconfig

The chain-length DTS property is now mandatory, so remove the
optional Kconfig for setting the length

Signed-off-by: Jamie McCrae <spam@helper3000.net>

simonguinot commented on include/zephyr/drivers/led_strip.h:
```diff
@@ -110,6 +111,16 @@ static inline int led_strip_update_rgb(const struct device *dev,
+			return -ERANGE;
+		}
+	}
+
```
You can now remove the redundant check in `tlc5971_update_rgb()`.

And maybe the `num_pixels_ok()` check can be removed in `ws2812_strip_update_rgb()` as well.

thedjnK replied:
Removed the tlc5971 check, kept the ws2812 check as it's checking the buffer in regards to number of channels also

simonguinot replied:
The size of the internal buffer size used by the `ws2812_spi` is set to `cfg->num_colors * pixel-chain-length * 8`. Since `->update_rgb()` only copies `cfg->num_colors` for each passed pixel, then verifying the chain length is enough to ensure that the buffer is big enough. So `num_pixels_ok()` can be removed.

thedjnK replied:
Removed

simonguinot commented on dts/bindings/led_strip/led-strip.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2019, Linaro Limited
```
In the commit message, maybe you mean "color ordering property" instead of "ordering property" ?

thedjnK replied:
Updated

simonguinot commented on boards/adafruit/itsybitsy/adafruit_itsybitsy_nrf52840.dts:
```diff
@@ -8,6 +8,7 @@
 #include <nordic/nrf52840_qiaa.dtsi>
```
I think this commit would be better earlier in the series, right after "dts: bindings: led_strip: Add common binding". This would reduce the risk of breaking a build in case of a bisection.

thedjnK replied:
Moved

simonguinot commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -177,6 +177,14 @@ LED Strip
+* The ``chain-length`` and ``color-mapping`` properties have been added to all LED strip bindings
+  and are now mandatory.
+
+* Added a new mandatory ``length`` function which returns the length (number of pixels) of an LED
```
Since LED begins with a consonant sound, shouldn't it be "of a LED" instead ?

kartben replied:
It's actually a vowel sound: ... an el-ee-dee ...

simonguinot replied:
You are right ! For some reason I had the french sound in mind :)

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -231,6 +240,7 @@ static const struct led_strip_driver_api ws2812_spi_api = {
 		.zero_frame = WS2812_SPI_ZERO_FRAME(idx),		 \
 		.num_colors = WS2812_NUM_COLORS(idx),			 \
 		.color_mapping = ws2812_spi_##idx##_color_mapping,	 \
+		.length = DT_INST_PROP(idx, chain_length),               \
```
Maybe `px_buf_size` can be removed ?

thedjnK replied:
That deals with number of colours too so is not the same as the length

simonguinot replied:
Well, `cfg->px_buf_size = cfg->num_colors * cfg->length * 8`. So maybe we can remove one of this variables to save a few bytes in memory ?

thedjnK replied:
Removed

simonguinot reviewed (commented):
Hi  @thedjnK,

Thank you very much for these improvements and cleanings.

See below for my comments.

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -67,20 +67,6 @@ static inline void ws2812_spi_ser(uint8_t buf[8], uint8_t color,
-	overflow = size_mul_overflow(num_pixels, cfg->num_colors * 8, &nbytes);
-	return !overflow && (nbytes <= cfg->px_buf_size);
-}
-
```
You should keep this function alive until `drivers: led_strip: Check length before updating LED strip`.

thedjnK replied:
Fixed

simonguinot reviewed (dismissed):
Thanks @thedjnK !

pdgendt commented on include/zephyr/drivers/led_strip.h:
```diff
@@ -79,58 +89,85 @@ typedef int (*led_api_update_channels)(const struct device *dev,
+	/* Allow for out-of-tree drivers that do not have this function for 2 Zephyr releases
+	 * until making it mandatory, function added after Zephyr 3.6
+	 */
+	if (api->length) {
```
Nit: 

```suggestion
	if (api->length != NULL) {
```

aescolar commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -212,6 +212,14 @@ LED Strip
 * The property ``in-gpios`` defined in :dtcompatible:`worldsemi,ws2812-gpio` has been
   renamed to ``gpios``.
 
+* The ``chain-length`` and ``color-mapping`` properties have been added to all LED strip bindings
```
@kartben shouldn't we have similar point in the release notes? (or at least something like "LED strips: bindings and API have been updated"

kartben replied:
yep

thedjnK replied:
I don't see what needs to be added, the changes are described here, they're in migration guide and not release notes because these are changes users need to be aware of, this is how it's documented to put things in the migration guide/release notes

aescolar reviewed (dismissed):
See comment https://github.com/zephyrproject-rtos/zephyr/pull/71630#discussion_r1591968358

thedjnK commented:
ping @soburi
